mixer/pkg/runtime/config/yaml/reader.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yaml
+
+import (
+	"fmt"
+
+	"strings"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/pkg/errors"
+)
+
+// Callback defines the callback made by the Reader.
+type Callback interface {
+	Handle(name string, val interface{}, field *descriptor.FieldDescriptorProto) (Callback, error)
+}
+
+// Reader visits all the nodes inside yaml in a BFS, and calls back with the observed information.
+type Reader interface {
+	// Walk crawls through the yaml and executes the Callback callback
+	Walk(data map[string]interface{}, msgName string, skipUnknown bool, handler Callback) error
+}
+
+type reader struct {
+	resolver Resolver
+}
+
+// NewReader returns a reader that parses yaml
+func NewReader(resolver Resolver) Reader {
+	return &reader{resolver: resolver}
+}
+
+func (r *reader) Walk(data map[string]interface{}, msgName string, skipUnknown bool, handler Callback) error {
+	message := r.resolver.FindMessage(msgName)
+	if message == nil {
+		return fmt.Errorf("cannot resolve message name '%s'", msgName)
+	}
+	for k, v := range data {
+		if fieldDescriptor := findFieldByName(message, k); fieldDescriptor == nil {
+			if skipUnknown {
+				continue
+			}
+			return fmt.Errorf("error when parsing config '%v' into message '%s'; unknown field '%s'", data,
```

Comment:
For reusability, I suggest not using the word "config" in this and other error messages. A this level, they're yaml documents. If we want to add the word "config" in the error messages, that should happen at a higher level.

Reply:
can you simplify this logic, something like
```go
fd := findFieldByName(message, k)
if fd == nil {
  if skipUnknown {
    continue
 }
 return Error()
}

return r.walk (.....)
```